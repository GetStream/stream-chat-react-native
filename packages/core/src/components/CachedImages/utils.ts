export const extractPathname = (uri: string | undefined) => {
  const parsedUrl = uri?.split('?')?.[0];
  // Extracts pathname. Need to use a regex cause url.parse is deprecated and RN doesnt have
  // URL constructor. Replace / with ___ to avoid creating folders
  return parsedUrl?.replace(/(https?:\/\/)?/, '').replace(/\//g, '___');
};
