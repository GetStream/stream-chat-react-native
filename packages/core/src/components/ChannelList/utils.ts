import type { Channel, StreamChat } from 'stream-chat';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

type MoveParameters<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  channels: Channel<At, Ch, Co, Ev, Me, Re, Us>[];
  cid: string;
};

export const moveChannelUp = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>({
  cid,
  channels = [],
}: MoveParameters<At, Ch, Co, Ev, Me, Re, Us>) => {
  // get channel from channels
  const index = channels.findIndex((c) => c.cid === cid);
  if (index <= 0) return channels;
  const channel = channels[index];

  // remove channel from current position and add to start
  channels.splice(index, 1);
  channels.unshift(channel);

  return [...channels];
};

type GetParameters<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  client: StreamChat<At, Ch, Co, Ev, Me, Re, Us>;
  id: string;
  type: string;
};

export const getChannel = async <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>({
  client,
  id,
  type,
}: GetParameters<At, Ch, Co, Ev, Me, Re, Us>) => {
  const channel = client.channel(type, id);
  await channel.watch();
  return channel;
};

export const DEFAULT_QUERY_CHANNELS_LIMIT = 10;
export const MAX_QUERY_CHANNELS_LIMIT = 30;
