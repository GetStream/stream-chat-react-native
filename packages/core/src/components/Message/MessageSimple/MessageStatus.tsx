import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

import {
  MessageContextValue,
  useMessageContext,
} from '../../../contexts/messageContext/MessageContext';
import { useTheme } from '../../../contexts/themeContext/ThemeContext';
import { Check } from '../../../icons/Check';
import { CheckAll } from '../../../icons/CheckAll';
import { Time } from '../../../icons/Time';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import { isMessageWithStylesReadByAndDateSeparator } from '../../MessageList/hooks/useMessageList';

const styles = StyleSheet.create({
  readByCount: {
    fontSize: 11,
    fontWeight: '700',
    paddingRight: 3,
  },
  statusContainer: {
    alignItems: 'flex-end',
    flexDirection: 'row',
    justifyContent: 'center',
    paddingRight: 3,
  },
});

export type MessageStatusPropsWithContext<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
> = Pick<MessageContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'message' | 'threadList'>;

const MessageStatusWithContext = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
>(
  props: MessageStatusPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { message, threadList } = props;

  const {
    theme: {
      colors: { accent_blue, grey_dark },
      messageSimple: {
        status: { checkAllIcon, checkIcon, readByCount, statusContainer, timeIcon },
      },
    },
  } = useTheme();

  if (message.status === 'sending') {
    return (
      <View style={[styles.statusContainer, statusContainer]} testID='sending-container'>
        <Time {...timeIcon} />
      </View>
    );
  }

  if (isMessageWithStylesReadByAndDateSeparator(message) && !threadList) {
    return (
      <View style={[styles.statusContainer, statusContainer]}>
        {typeof message.readBy === 'number' ? (
          <Text
            style={[styles.readByCount, { color: accent_blue }, readByCount]}
            testID='read-by-container'
          >
            {message.readBy}
          </Text>
        ) : null}
        {typeof message.readBy === 'number' || message.readBy === true ? (
          <CheckAll pathFill={accent_blue} {...checkAllIcon} />
        ) : (
          <Check pathFill={grey_dark} {...checkIcon} />
        )}
      </View>
    );
  }

  if (message.status === 'received' && message.type !== 'ephemeral' && !threadList) {
    return (
      <View style={[styles.statusContainer, statusContainer]} testID='delivered-container'>
        <Check pathFill={grey_dark} {...checkIcon} />
      </View>
    );
  }

  return null;
};

const areEqual = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  prevProps: MessageStatusPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
  nextProps: MessageStatusPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { message: prevMessage, threadList: prevThreadList } = prevProps;
  const { message: nextMessage, threadList: nextThreadList } = nextProps;

  const threadListEqual = prevThreadList === nextThreadList;
  if (!threadListEqual) return false;

  const messageEqual =
    prevMessage.status === nextMessage.status &&
    prevMessage.type === nextMessage.type &&
    (isMessageWithStylesReadByAndDateSeparator(prevMessage) && prevMessage.readBy) ===
      (isMessageWithStylesReadByAndDateSeparator(nextMessage) && nextMessage.readBy);
  if (!messageEqual) return false;

  return true;
};

const MemoizedMessageStatus = React.memo(
  MessageStatusWithContext,
  areEqual,
) as typeof MessageStatusWithContext;

export type MessageStatusProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
> = Partial<MessageStatusPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>>;

export const MessageStatus = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
>(
  props: MessageStatusProps<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { message, threadList } = useMessageContext<At, Ch, Co, Ev, Me, Re, Us>();

  return <MemoizedMessageStatus {...{ message, threadList }} {...props} />;
};

MessageStatus.displayName = 'MessageStatus{messageSimple{status}}';
