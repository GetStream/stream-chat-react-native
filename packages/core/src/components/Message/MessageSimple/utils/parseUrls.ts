const protocol = '[a-z][\\w\\.+-]+:\\/\\/';
const hostName = '[w]{3,}\\.';
const domainName = '[\\w_-]+((?:\\.[\\w_-]+)+)';
const path = '[\\w\\p{Letter}()$/\\._-]+';
const queryString = '\\?[\\w\\p{Letter}=&+~_-]+';
const fragment = '#[\\w\\p{Letter}#?=&+~_-]+';
const port = ':[\\d]{2,5}';
const emailAllowedCharacters = '[\\w+\\.~$_-]+';

/**
 * Don't match the following pattern if the previous string
 * matches `[anything](` as we don't want to interfere
 * with links that already have been formatted as markdown
 * */
const notWithinLink = `(?<!\\[.+\\]\\(.+)`;

/**
 * Match any email address, with and without `mailto:`
 * */
const emailPattern = `(?:mailto:)?(?<emailDisplay>(?:${emailAllowedCharacters})@(?:${domainName}))`;
/**
 * Match links decided by having a protocol, allows for example
 * https://localhost to be recognized as a link
 * */
const protocolPrefixedLinkPattern = `(?<pplProtocol>${protocol})(?<pplDisplay>${notWithinLink}(?:${path})?(?:${queryString})?(?:${fragment})?)`;
/**
 * Match as much as possible of a fqdn, at least with a domain name formed as
 * something.tld
 * */
const fqdnPattern = `(?<fqdnProtocol>${protocol})?(?:${hostName})?(?<fqdnDisplay>${notWithinLink}(?:${domainName})(?:${port})?(?:${path})?(?:${queryString})?(?:${fragment})?)`;

const linkRegexp = new RegExp(
  `${emailPattern}|${fqdnPattern}|${protocolPrefixedLinkPattern}`,
  'giu',
);

interface Link {
  encoded: string;
  protocol: string;
  raw: string;
}

export const parseUrlsFromText = (input: string): Link[] => {
  let matches;

  const results: Array<{
    encoded: string;
    protocol: string;
    raw: string;
  }> = [];

  while ((matches = linkRegexp.exec(input)) !== null) {
    results.push(extractFromMatchGroups(matches[0], matches.groups ?? {}));
  }

  return results;
};

/**
 * We can't use the same names for multiple groups in a regexp,
 * this helper accepts all group names and builds a `Link`
 * object from it.
 * */
const extractFromMatchGroups = (
  rawLink: string,
  {
    emailDisplay = null,
    fqdnDisplay = null,
    fqdnProtocol = null,
    pplDisplay = null,
    pplProtocol = null,
  },
): Link => {
  if (emailDisplay !== null) {
    return {
      encoded: encodeURI(emailDisplay),
      protocol: 'mailto:',
      raw: rawLink,
    };
  }

  return {
    encoded: encodeURI(fqdnDisplay ?? pplDisplay ?? ''),
    protocol: fqdnProtocol ?? pplProtocol ?? '',
    raw: rawLink,
  };
};
