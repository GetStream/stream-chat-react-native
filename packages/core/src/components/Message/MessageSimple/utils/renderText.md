[src/components/Message/MessageSimple/utils/renderText.ts](https://github.com/GetStream/stream-chat-react-native/blob/master/src/components/Message/MessageSimple/utils/renderText.ts)

Function that handles markdown and parsing for the text of message.

Internally it uses [react-native-simple-markdown](https://github.com/GetStream/react-native-simple-markdown#styles-1) to handle markdown.

It accepts following params:

- markdownRules { object } - <https://github.com/GetStream/react-native-simple-markdown#rules>
- markdownStyles { object } - <https://github.com/GetStream/react-native-simple-markdown#styles-1>
- message { object } Message object
