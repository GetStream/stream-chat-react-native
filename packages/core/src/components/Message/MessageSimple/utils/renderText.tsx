import React from 'react';
import { GestureResponderEvent, Linking, Text } from 'react-native';
import anchorme from 'anchorme';
import truncate from 'lodash/truncate';
// @ts-expect-error
import Markdown from 'react-native-markdown-package';
import {
  DefaultRules,
  defaultRules,
  MatchFunction,
  ParseFunction,
  parseInline,
  ReactNodeOutput,
} from 'simple-markdown';

import type { MessageType } from '../../../MessageList/hooks/useMessageList';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../../types/types';
import type { MessageContextValue } from '../../../../contexts/messageContext/MessageContext';
import type { Colors, MarkdownStyle } from '../../../../contexts/themeContext/utils/theme';

const defaultMarkdownStyles: MarkdownStyle = {
  inlineCode: {
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5,
  },
  list: {
    marginBottom: 8,
    marginTop: 8,
  },
  listItemText: {
    flex: 0,
  },
  mentions: {
    fontWeight: '700',
  },
  // unfortunately marginVertical doesn't override the defaults for these within the 3rd party lib
  paragraph: {
    marginBottom: 8,
    marginTop: 8,
  },
  paragraphCenter: {
    marginBottom: 8,
    marginTop: 8,
  },
  paragraphWithImage: {
    marginBottom: 8,
    marginTop: 8,
  },
};

const parse: ParseFunction = (capture, parse, state) => ({
  content: parseInline(parse, capture[0], state),
});

export type MarkdownRules = Partial<DefaultRules>;

export type RenderTextParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = Partial<
  Pick<MessageContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'onLongPress' | 'onPress' | 'preventPress'>
> & {
  colors: typeof Colors;
  message: MessageType<At, Ch, Co, Ev, Me, Re, Us>;
  markdownRules?: MarkdownRules;
  markdownStyles?: MarkdownStyle;
  messageOverlay?: boolean;
  onLink?: (url: string) => Promise<void>;
  onlyEmojis?: boolean;
};

export const renderText = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  params: RenderTextParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    colors,
    markdownRules,
    markdownStyles,
    message,
    messageOverlay,
    onLink: onLinkParams,
    onLongPress: onLongPressParam,
    onlyEmojis,
    onPress: onPressParam,
    preventPress,
  } = params;

  // take the @ mentions and turn them into markdown?
  // translate links
  const { mentioned_users, text } = message;

  if (!text) return null;

  let newText = text.trim();
  const urls = anchorme(newText, {
    list: true,
  });

  for (const urlInfo of urls) {
    const displayLink = truncate(urlInfo.encoded.replace(/^(www\.)/, ''), {
      length: 200,
      omission: '...',
    });
    const markdown = `[${displayLink}](${urlInfo.protocol}${urlInfo.encoded})`;
    newText = newText.replace(urlInfo.raw, markdown);
  }

  newText = newText.replace(/[<&"'>]/g, '\\$&');
  const styles: MarkdownStyle = {
    ...defaultMarkdownStyles,
    ...markdownStyles,
    autolink: {
      ...defaultMarkdownStyles.autolink,
      color: colors.accent_blue,
      ...markdownStyles?.autolink,
    },
    inlineCode: {
      ...defaultMarkdownStyles.inlineCode,
      backgroundColor: colors.white_smoke,
      borderColor: colors.grey_gainsboro,
      color: colors.accent_red,
      ...markdownStyles?.inlineCode,
    },
    mentions: {
      ...defaultMarkdownStyles.mentions,
      color: colors.accent_blue,
      ...markdownStyles?.mentions,
    },
    text: {
      ...defaultMarkdownStyles.text,
      color: colors.black,
      ...markdownStyles?.text,
    },
  };

  const onLink = (url: string) =>
    onLinkParams
      ? onLinkParams(url)
      : Linking.canOpenURL(url).then((canOpenUrl) => canOpenUrl && Linking.openURL(url));

  const react: ReactNodeOutput = (node, output, { ...state }) => {
    const onPress = (event: GestureResponderEvent) => {
      if (!preventPress && onPressParam) {
        onPressParam({
          defaultHandler: () => onLink(node.target),
          emitter: 'textLink',
          event,
        });
      }
    };

    const onLongPress = (event: GestureResponderEvent) => {
      if (!preventPress && onLongPressParam) {
        onLongPressParam({
          emitter: 'textLink',
          event,
        });
      }
    };

    state.withinLink = true;
    const link = React.createElement(
      Text,
      {
        key: state.key,
        onLongPress,
        onPress,
        style: styles.autolink,
        suppressHighlighting: true,
      },
      output(node.content, state),
    );
    state.withinLink = false;
    return link;
  };

  const mentionedUsers = Array.isArray(mentioned_users)
    ? mentioned_users.reduce((acc, cur) => {
        const userName = cur.name || cur.id || '';
        if (userName) {
          acc += `${acc.length ? '|' : ''}@${userName}`;
        }
        return acc;
      }, '')
    : '';

  const regEx = new RegExp(`^\\B(${mentionedUsers})`, 'g');
  const match: MatchFunction = (source) => regEx.exec(source);

  const mentionsReact: ReactNodeOutput = (node, output, { ...state }) => {
    const onPress = (event: GestureResponderEvent) => {
      if (!preventPress && onPressParam) {
        onPressParam({
          emitter: 'textMention',
          event,
        });
      }
    };

    const onLongPress = (event: GestureResponderEvent) => {
      if (!preventPress && onLongPressParam) {
        onLongPressParam({
          emitter: 'textMention',
          event,
        });
      }
    };

    return React.createElement(
      Text,
      {
        key: state.key,
        onLongPress,
        onPress,
        style: styles.mentions,
      },
      Array.isArray(node.content) ? node.content[0]?.content || '' : output(node.content, state),
    );
  };

  const customRules = {
    link: { react },
    // we have no react rendering support for reflinks
    reflink: { match: () => null },
    ...(mentionedUsers
      ? {
          mentions: {
            match,
            order: defaultRules.text.order - 0.5,
            parse,
            react: mentionsReact,
          },
        }
      : {}),
  };

  return (
    <Markdown
      key={`${JSON.stringify(mentioned_users)}-${onlyEmojis}-${
        messageOverlay ? JSON.stringify(markdownStyles) : undefined
      }`}
      onLink={onLink}
      rules={{
        ...customRules,
        ...markdownRules,
      }}
      styles={styles}
    >
      {newText}
    </Markdown>
  );
};
