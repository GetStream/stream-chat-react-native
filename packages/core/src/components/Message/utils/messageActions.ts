import type { MessageActionType } from '../../MessageOverlay/MessageActionListItem';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import type { MessageContextValue } from '../../../contexts/messageContext/MessageContext';
import type { MessagesContextValue } from '../../../contexts/messagesContext/MessagesContext';

export type MessageActionsProps = {
  blockUser: MessageActionType;
  copyMessage: MessageActionType;
  deleteMessage: MessageActionType;
  editMessage: MessageActionType;
  flagMessage: MessageActionType;
  muteUser: MessageActionType;
  pinMessage: MessageActionType;
  quotedReply: MessageActionType;
  retry: MessageActionType;
  threadReply: MessageActionType;
  unpinMessage: MessageActionType;
};

export const messageActions = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>({
  blockUser,
  canModifyMessage,
  copyMessage,
  deleteMessage,
  editMessage,
  error,
  flagMessage,
  isMyMessage,
  isThreadMessage,
  message,
  messageReactions,
  mutesEnabled,
  muteUser,
  pinMessage,
  pinMessageEnabled,
  quotedRepliesEnabled,
  quotedReply,
  retry,
  threadRepliesEnabled,
  threadReply,
  unpinMessage,
}: {
  canModifyMessage: boolean;
  error: boolean | Error;
  isThreadMessage: boolean;
  messageReactions: boolean;
} & MessageActionsProps &
  Pick<MessageContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'message' | 'isMyMessage'> &
  Pick<
    MessagesContextValue<At, Ch, Co, Ev, Me, Re, Us>,
    'mutesEnabled' | 'quotedRepliesEnabled' | 'pinMessageEnabled' | 'threadRepliesEnabled'
  >): Array<MessageActionType | null> | undefined => {
  if (messageReactions) {
    return undefined;
  }

  const actions: Array<MessageActionType | null> = [];

  if (error && isMyMessage) {
    actions.push(retry);
  }

  if (quotedRepliesEnabled && !isThreadMessage && !error) {
    actions.push(quotedReply);
  }

  if (threadRepliesEnabled && !isThreadMessage && !error) {
    actions.push(threadReply);
  }

  if (canModifyMessage) {
    actions.push(editMessage);
  }

  if (message.text && !error) {
    actions.push(copyMessage);
  }

  if (!isMyMessage) {
    actions.push(flagMessage);
  }

  if (pinMessageEnabled && !message.pinned) {
    actions.push(pinMessage);
  }

  if (pinMessageEnabled && message.pinned) {
    actions.push(unpinMessage);
  }

  if (mutesEnabled && !isMyMessage) {
    actions.push(muteUser);
  }

  if (!isMyMessage && canModifyMessage) {
    actions.push(blockUser);
  }

  if (canModifyMessage) {
    actions.push(deleteMessage);
  }

  return actions;
};
