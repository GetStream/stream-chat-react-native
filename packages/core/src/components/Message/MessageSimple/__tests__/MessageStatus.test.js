import React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react-native';

import { MessageStatus } from '../MessageStatus';

import { Chat } from '../../../Chat/Chat';

import { Streami18n } from '../../../../utils/Streami18n';
import { generateMessage } from '../../../../mock-builders/generator/message';
import { generateStaticUser, generateUser } from '../../../../mock-builders/generator/user';
import { getTestClientWithUser } from '../../../../mock-builders/mock';

let chatClient;
let id;
let i18nInstance;

describe('MessageStatus', () => {
  beforeAll(async () => {
    id = 'testID';
    chatClient = await getTestClientWithUser({ id });
    i18nInstance = new Streami18n();
  });
  afterEach(cleanup);

  it('should render message status with delivered container', async () => {
    const user = generateUser();
    const message = generateMessage({ user });

    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus lastReceivedId={message.id} message={{ ...message, status: 'received' }} />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('delivered-container')).toBeTruthy();
    });
  });

  it('should render message status with read by container', async () => {
    const user = generateUser();
    const message = generateMessage({ readBy: 2, user });

    const { getByTestId, getByText, rerender, toJSON } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus lastReceivedId={message.id} message={message} />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('read-by-container')).toBeTruthy();
      expect(getByText(message.readBy.toString())).toBeTruthy();
    });

    const staticUser = generateStaticUser(0);
    const staticMessage = generateMessage({ readBy: 2, staticUser });

    rerender(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus lastReceivedId={staticMessage.id} message={staticMessage} />
      </Chat>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
      expect(getByTestId('read-by-container')).toBeTruthy();
      expect(getByText(staticMessage.readBy.toString())).toBeTruthy();
    });
  });

  it('should render message status with sending container', async () => {
    const user = generateUser();
    const message = generateMessage({ user });

    const { getByTestId } = render(
      <Chat client={chatClient} i18nInstance={i18nInstance}>
        <MessageStatus message={{ ...message, status: 'sending' }} />
      </Chat>,
    );

    await waitFor(() => {
      expect(getByTestId('sending-container')).toBeTruthy();
    });
  });
});
