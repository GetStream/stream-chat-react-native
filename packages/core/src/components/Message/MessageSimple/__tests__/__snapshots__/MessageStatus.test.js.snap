// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageStatus should render message status with read by container 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingRight": 3,
      },
      Object {},
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 11,
          "fontWeight": "700",
          "paddingRight": 3,
        },
        Object {
          "color": "#005FFF",
        },
        Object {},
      ]
    }
    testID="read-by-container"
  >
    2
  </Text>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={16}
    bbWidth={16}
    focusable={false}
    height={16}
    meetOrSlice={0}
    minX={0}
    minY={0}
    pathFill="#005FFF"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 16,
          "width": 16,
        },
      ]
    }
    vbHeight={24}
    vbWidth={24}
    width={16}
  >
    <RNSVGGroup>
      <RNSVGPath
        clipRule={0}
        d="M2.293 11.293a1 1 0 011.414 0l4 4a1 1 0 11-1.414 1.414l-4-4a1 1 0 010-1.414zM9.293 12.293a1 1 0 011.414 0l3 3a1 1 0 01-1.414 1.414l-3-3a1 1 0 010-1.414z"
        fill={4278214655}
        fillRule={0}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
      <RNSVGPath
        clipRule={0}
        d="M15.707 7.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414-1.414l8-8a1 1 0 011.414 0zM21.707 7.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414-1.414l8-8a1 1 0 011.414 0z"
        fill={4278214655}
        fillRule={0}
        propList={
          Array [
            "fill",
            "fillRule",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
